src/core/experimental/refcount.d

```diff
@@ -0,0 +1,782 @@
+/**
+  This module provides a composable reference count implementation in the form
+  of `__RefCount`.
+*/
+module core.experimental.refcount;
+
+/**
+ * A qualified reference counted struct that is intended to be composed by user
+ * defined types that desire to implement manual memory management by means of
+ * reference counting. Note to user: The internal implementation uses malloc/free.
+ *
+ * `__RefCount` was designed to be composed as a field inside the user defined type.
+ * The user is responsible to initialize the `__RefCount` in the constructor of his
+ * type. The user will call the `isUnique()` method to decide if this is the last
+ * reference to his type so he can safely deallocate his own managed memory.
+ *
+ * `Important`: the `__RefCount` must be initialized through a call to the
+ * constructor before being used.
+ */
+struct __RefCount
+{
+    import core.atomic : atomicOp;
+
+    alias CounterType = uint;
+    private CounterType* rc = null;
+
+    /**
+     * Return a boolean value denoting if this can be used in a shared context.
+     *
+     * Returns:
+     *      `true` if this started off as an immutable object; `false` otherwise.
+     *
+     * Complexity:
+     *      $(BIGOH 1).
+     */
+    @nogc nothrow pure @safe scope
+    bool isShared() const
+    {
+        // Faster than ((cast(size_t) rc) % 8) == 0;
+        return !((cast(size_t) rc) & 7);
+    }
+
+    /*
+     * Perform `rc op val` operation. Use atomics if this is shared.
+     *
+     * Returns:
+     *      The result of the operation.
+     */
+    @nogc nothrow pure @trusted scope
+    private CounterType rcOp(string op)(CounterType val) const
+    {
+        if (isShared())
+        {
+            return cast(CounterType)(atomicOp!op(*(cast(shared CounterType*) rc), val));
+        }
+        else
+        {
+            mixin("return cast(CounterType)(*(cast(CounterType*) rc)" ~ op ~ "val);");
+        }
+    }
+
+    /**
+     * Creates a new `__RefCount` instance. It's memory is internally managed with
+     * malloc/free.
```

Comment:
So no other allocators then? I know this is a druntime PR so it wouldn't be able to use std.experimental.allocator, but then I wonder why not put this in phobos.

Reply:
It's placed in druntime as it might require some compiler "help" in the future. For now, all the functions that are called for a dec ref return a `void*`, ex `void* delRef() const`. This is so in the context of an `immutable` object the `pure const delRef` call won't be elided. Currently there is no magic involved, but If, in the future, the compiler will see that the returned value is never used and it decides to elide, we might need a helping hand.

Reply:
> call won't be elided. 

Then I'd be making sure to thoroughly with LDC and GDC. Also in the case of dead object elimination (as opposed to just dead call) that is the desired outcome.

Reply:
> Also in the case of dead object elimination (as opposed to just dead call) that is the desired outcome.

Good point. I'd assume that in this case, since the object is "dead" it will elide the calls regardless of any other considerations, as the calls are made through the object

Reply:
Any such compiler help would require a DIP, no?

Reply:
> Any such compiler help would require a DIP, no?

I believe so

Reply:
>So no other allocators then? I know this is a druntime PR so it wouldn't be able to use std.experimental.allocator, but then I wonder why not put this in phobos.

We wanted to get a minimal version going because we have these huge problems with purity and immutable etc. Worrying about custom memory allocation is additional distraction. Too many holes in the dam, too few fingers.

We must put this in druntime because that's where the refcounted slice needs to be (i.e. a slice that is just like `T[]` but manages its own memory).